doc/source-repositories.rst

```diff
@@ -0,0 +1,224 @@
+Source Repositories
+===================
+
+Source repositories are a way to specify where to find the source code for a
+package, both for :ref:`package authors<pkg-authors>` and maintainers and for
+:ref:`package consumers<pkg-consumers>`.
+
+A relatively structured set of version control system (VCS) fields, that vary
+depending on the :ref:`VCS kind<vcs-kind>`, enables cabal commands and other
+tools to interpret and make effective use of this information.
+
+- ``source-repository`` says where to find the source for a package description
+- ``source-repository-package`` locates source for project dependencies
+
+.. _pkg-authors:
+
+As a Package Author
+-------------------
+
+If you are authoring or maintaining a package, you will be dealing with
+``source-repository``, a package description stanza that typically specifies
+where upstream development of the package is happening.  A package published to
+hackage will be displayed with a link to the source repository if a
+``source-repository`` stanza is present in the package description.
+
+.. _pkg-consumers:
+
+As a Package Consumer
+---------------------
+
+When working on a project you may need one or more ``source-repository-package``
+stanzas if taking dependencies via a VCS such as git.  This is an alternative
+way of taking dependencies, instead of getting them from a package repository
+such as hackage. It may be the only practical way of taking dependencies on
+packages not published to hackage. For a package on hackage, a
+``source-repository`` stanza is not needed but if used, the downloaded source
+will become the only version of the package available to the project.
+
+A dependency taken this way, effectively adds a local package to the project
+much like listing the package's source directory in the packages field, except
+that the source code is downloaded by cabal using the version control system.
+Cabal commands that need the source code will do this as needed, commands like
+``cabal build`` or ``cabal freeze``. This is also the easiest way to work with a
+fork, much easier than using a git submodule.
+
+.. _source-repository-fields:
+
+Source Repository Fields
+------------------------
+..
+  data SourceRepo = SourceRepo
+    { repoKind :: RepoKind
+    , repoType :: Maybe RepoType
+    , repoLocation :: Maybe String
+    , repoModule :: Maybe String
+    , repoBranch :: Maybe String
+    , repoTag :: Maybe String
+    , repoSubdir :: Maybe FilePath
+    }
```

Comment:
I would advise against add this Haskell code, because as a user of cabal i'm not that interested in the underlying implementation. If you insist that this information is useful, maybe a link to the actual implementation from the refencre section would be good alternative.

Reply:
That code snippet is an RST comment and not rendered.